File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/hashx/fletcher32.go
package hashx

// Fletcher32 computes the fletcher32 checksum over raw bytes.
func Fletcher32(data []byte) uint64 {
	low := uint32(0)
	high := uint32(0)
	for i := 0; i < len(data); i += 2 {
		word := uint32(data[i])
		if i+1 < len(data) {
			word |= uint32(data[i+1]) << 8
		}
		low = (low + word) % 65535
		high = (high + low) % 65535
	}
	return uint64((high << 16) | low)
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/fletcher32.cpp
#include "freight/hashes.h"

namespace freight {

// fletcher32 over raw bytes.
uint64_t hash_fletcher32(const std::string& data) {
  uint32_t low = 0u;
  uint32_t high = 0u;
  for (size_t i = 0; i < data.size(); i += 2) {
    uint32_t word = static_cast<uint8_t>(data[i]);
    if (i + 1 < data.size()) {
      word |= static_cast<uint32_t>(static_cast<uint8_t>(data[i + 1])) << 8;
    }
    low = (low + word) % 65535u;
    high = (high + low) % 65535u;
  }
  return static_cast<uint64_t>((high << 16) | low);
}

}  // namespace freight
